### PR TITLE
Enable Observability Control Flow Recording

### PR DESCRIPTION
**Pull request by sachiniSam**
## Purpose
With this PR, we could enhance the observability by enabling the control flow recording feature. Users will be able to detect the exact request path as well as the time taken between each construct.

## Approach
The observability instrumentation has been carried out at the BIR level. current instrumentation applies for the packages with `entry points`. The JI calls for the Basic blocks are added based on the positions of `instructions` and `terminators` while the priority has been given for the `instructions` position.

These `checkpoints` added could be viewed using Jaeger under the Logs with the `moduleID` and `positionID` as a key,value pair

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java [hunk 1]
```diff
@@ -170,6 +193,65 @@ void instrumentPackage(BIRPackage pkg) {
...
+                //Fix error entries in the error entry table
+                fixErrorTable(func, startBB, newBB);
+                i += 1; // Number of inserted BBs
+            }
+            i += 1;
+        }
+    }
+
+    /**
+     * Inject checkpoint JI method call to a basic block.
```
Comment (nadundesilva) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java:233]: Should have an empty line between description and params.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java [hunk 2]
```diff
@@ -170,6 +193,65 @@ void instrumentPackage(BIRPackage pkg) {
...
+     * @param startBB The basic block to which the checkpoint call should be injected
+     * @param pkg The package the invocation belongs to
+     * @param desugaredInsPosition The source code position of the invocation
+     */
+    private void injectCheckpointCall(BIRBasicBlock startBB, BIRPackage pkg, Location desugaredInsPosition) {
```
Comment (nadundesilva) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java:238]: Shall we rename `desugaredInsPosition` to `originalInsPosition` ? Both are the same here, but we switch names halfway in this function method body.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java [hunk 3]
```diff
@@ -170,6 +193,65 @@ void instrumentPackage(BIRPackage pkg) {
...
+        String pkgId = generatePackageId(pkg);
+        String position = generatePositionId(desugaredInsPosition);
+
+        BIROperand pkgOperand = generateGlobalConstantOperand(pkg, symbolTable.stringType, pkgId);
+        BIROperand originalInsPosOperand = generateGlobalConstantOperand(pkg, symbolTable.stringType, position);
+
+        JIMethodCall observeStartCallTerminator = new JIMethodCall(null);
```
Comment (nadundesilva) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java:245]: Let's rename `observeStartCallTerminator` to `recordCheckPointCallTerminator`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java [hunk 4]
```diff
@@ -108,6 +109,9 @@
     private static final String FUNC_BODY_INSTRUMENTATION_TYPE = "funcBody";
     private static final Location COMPILE_TIME_CONST_POS =
             new BLangDiagnosticLocation(null, -1, -1, -1, -1);
+    private static final String INIT_FUNC = ".<init>";
+    private static final String START_FUNC = ".<start>";
+    private static final String STOP_FUNC = ".<stop>";
```
[the remark this answers is not in the document]
Comment (sachiniSam) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java:114]: Using the `INIT_FUNCTION_SUFFIX` defined in `BLangConstants` creates a circular dependacy between ballerina-core.main and ballerina-lang.main. Therefore i redefined the  `INIT_FUNCTION_SUFFIX` constants like in `MethodGenUtils` and `ErrorValue`

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/tracing/MainFunctionTestCase.java [hunk 5]
```diff
@@ -73,6 +96,7 @@ public void testMainMethod() throws Exception {
                     new AbstractMap.SimpleEntry<>("src.entry_point.main", "true"),
                     new AbstractMap.SimpleEntry<>("function", "main")
             ));
+           Assert.assertEquals(span.getCheckpoints(), expectedCheckpoints);
```
Comment (nadundesilva) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/tracing/MainFunctionTestCase.java:99]: Minor alignment issue

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/tracing/MainFunctionTestCase.java [hunk 6]
```diff
@@ -48,6 +48,29 @@ public void testMainMethod() throws Exception {
...
+                new BMockSpan.BMockSpanEvent(moduleID, ":0:0"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":54:31"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":55:5"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":55:5"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":56:11"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":56:5"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":56:5"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":57:1"),
+                new BMockSpan.BMockSpanEvent(moduleID, FILE_NAME + ":29:9")
+                );
```
Comment (nadundesilva) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/tracing/MainFunctionTestCase.java:73]: Should align with `final` keyword

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 7]
```diff
@@ -1526,9 +1527,18 @@ public void visit(BLangIf astIfStmt) {
 
             // If a terminator statement has not been set for the else-block then just add it.
             if (this.env.enclBB.terminator == null) {
-                this.env.enclBB.terminator = new BIRTerminator.GOTO(null, nextBB);
+                if (astIfStmt.elseStmt.pos != null) {
+                    Location newLocation = new BLangDiagnosticLocation(
+                            astIfStmt.elseStmt.pos.lineRange().filePath(),
+                            astIfStmt.elseStmt.pos.lineRange().endLine().line(),
+                            astIfStmt.elseStmt.pos.lineRange().endLine().line(),
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:1534]: should this be start line ?

File: misc/tracing-extensions/modules/ballerina-choreo-extension/src/main/java/org/ballerinalang/observe/trace/extension/choreo/model/ChoreoTraceSpan.java [hunk 8]
```diff
@@ -112,4 +119,31 @@ public Type getRefType() {
             FOLLOWS_FROM
         }
     }
+
+    /**
+    * Trace Span Event.
+    */
+    public static class SpanEvent {
```
Comment (manuranga) [inline misc/tracing-extensions/modules/ballerina-choreo-extension/src/main/java/org/ballerinalang/observe/trace/extension/choreo/model/ChoreoTraceSpan.java:126]: let's make this a top level.

Comment (sachiniSam) [reply misc/tracing-extensions/modules/ballerina-choreo-extension/src/main/java/org/ballerinalang/observe/trace/extension/choreo/model/ChoreoTraceSpan.java:126]: fixed with the #27702 PR

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/ObserveUtils.java [hunk 9]
```diff
@@ -124,6 +125,37 @@ public static void startResourceObservation(Environment env, BString serviceName
...
+        if (span == null) {
+            return;
+        }
+
+        // Adding Position and Module ID to the Jaeger Span
+        Map<String, String> eventAttributes = new HashMap<>(2);
+        eventAttributes.put(TAG_KEY_MODULE, pkg.getValue());
+        eventAttributes.put(TAG_KEY_INVOCATION_POSITION, position.getValue());
+
+        HashMap<String, Object> events = new HashMap<>(1);
```
Comment (tishan89) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/ObserveUtils.java:154]: Declaration should be Map

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/TestReportTest.java [hunk 10]
```diff
@@ -62,7 +62,7 @@ public void testWarningForReportTools() throws BallerinaTestException, IOExcepti
         clientLeecher.waitForText(60000);
     }
 
-    @Test ()
+    @Test (enabled = false) // TODO: Enable the test case once the mismatch on line coverage is fixed. issue:#27524
```
Comment (tishan89) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/TestReportTest.java:65]: is this enabled?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/ObserveUtils.java (hunk 9 above)
Comment (sachiniSam) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/ObserveUtils.java:154]: Usage of a HashMap has been removed with the migration to Opentelemetry ->https://github.com/ballerina-platform/ballerina-lang/commit/4b8659bb0bc82db1593ba1c008e25937dd9c075d

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/TestReportTest.java (hunk 10 above)
Comment (sachiniSam) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/TestReportTest.java:65]: yes, it has been enabled with the pr -> https://github.com/ballerina-platform/ballerina-lang/commit/8f5e06ceb3cd62f446b2bc4fdd5f0a29a73c277c

**Pull request merged**